### PR TITLE
Pyic 3750 nino claim in identity

### PR DESCRIPTION
**Pull request by DanCorderIPV**
### What changed

If a successful NINO VC exists for a user when generating their UserIdentity a new claim containing the users NINO will be added to the UserIdentity 

### Why did it change

So the RPs can see the user's NINO

### Issue tracking

- [PYIC-3750](https://govukverify.atlassian.net/browse/PYIC-3750)

I have tested journeys both with and without the NINO CRI. When the NINO CRI isn't part of the journey the generated UserIdentity has no `https://vocab.account.gov.uk/v1/socialSecurityRecord` section at all. When the NINO CRI is in the journey the `https://vocab.account.gov.uk/v1/socialSecurityRecord` property exists and has the correct contents.

As the NINO CRI is behind a feature flag already this change shouldn't need it's own flag.

Also both the orchestrator and SPoT teams believe that even if the new claim was present it wouldn't cause them any issues.

[PYIC-3750]: https://govukverify.atlassian.net/browse/PYIC-3750?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: libs/user-identity-service/src/main/java/uk/gov/di/ipv/core/library/service/UserIdentityService.java [hunk 1]
```diff
@@ -48,11 +48,7 @@
 import static uk.gov.di.ipv.core.library.config.ConfigurationVariable.CORE_VTM_CLAIM;
 import static uk.gov.di.ipv.core.library.config.CoreFeatureFlag.EXIT_CODES;
 import static uk.gov.di.ipv.core.library.config.EnvironmentVariable.USER_ISSUED_CREDENTIALS_TABLE_NAME;
-import static uk.gov.di.ipv.core.library.domain.CriConstants.ADDRESS_CRI;
-import static uk.gov.di.ipv.core.library.domain.CriConstants.BAV_CRI;
-import static uk.gov.di.ipv.core.library.domain.CriConstants.DCMAW_CRI;
-import static uk.gov.di.ipv.core.library.domain.CriConstants.DRIVING_LICENCE_CRI;
-import static uk.gov.di.ipv.core.library.domain.CriConstants.PASSPORT_CRI;
+import static uk.gov.di.ipv.core.library.domain.CriConstants.*;
```
Comment (Joe-Edwards-GDS) [inline libs/user-identity-service/src/main/java/uk/gov/di/ipv/core/library/service/UserIdentityService.java:51]: It looks like elsewhere we're avoiding wildcard imports in favour of explicitly specifying them - might be worth keeping that consistent?

File: libs/user-identity-service/src/main/java/uk/gov/di/ipv/core/library/service/UserIdentityService.java [hunk 2]
```diff
@@ -333,131 +335,166 @@ public Optional<IdentityClaim> findIdentityClaim(List<VcStoreItem> vcStoreItems)
...
+                            SignedJWT.parse(ninoCredentialItem.getCredential())
+                                    .getPayload()
+                                    .toString())
+                    .path(VC_CLAIM)
+                    .path(VC_CREDENTIAL_SUBJECT)
+                    .path(criName);
+        } catch (JsonProcessingException | ParseException e) {
+            LOGGER.error("{}: '{}'", errorLog, e.getMessage());
+            throw new HttpResponseExceptionWithErrorBody(
+                    HttpStatus.SC_INTERNAL_SERVER_ERROR, errorResponse);
```
Comment (Joe-Edwards-GDS) [inline libs/user-identity-service/src/main/java/uk/gov/di/ipv/core/library/service/UserIdentityService.java:492]: I'm not sure if there's a sensible refactoring, but this is looking very similar to the `getVCClaimNode` method that @lwebb-iig modified recently.

Do you think there's scope to handle these more uniformly?

File: libs/user-identity-service/src/main/java/uk/gov/di/ipv/core/library/service/UserIdentityService.java [hunk 3]
```diff
@@ -333,131 +335,166 @@ public Optional<IdentityClaim> findIdentityClaim(List<VcStoreItem> vcStoreItems)
...
+    private Optional<VcStoreItem> findStoreItem(
+            List<String> criNames, List<VcStoreItem> vcStoreItems) {
+        return vcStoreItems.stream()
+                .filter(credential -> criNames.contains(credential.getCredentialIssuer()))
+                .findFirst();
+    }
+
+    private JsonNode extractCriNodeFromCredential(
+            String criName,
+            VcStoreItem ninoCredentialItem,
```
Comment (Joe-Edwards-GDS) [inline libs/user-identity-service/src/main/java/uk/gov/di/ipv/core/library/service/UserIdentityService.java]: should just be `credentialItem`

File: libs/user-identity-service/src/main/java/uk/gov/di/ipv/core/library/service/UserIdentityService.java (hunk 1 above)
Comment (Joe-Edwards-GDS) [reply libs/user-identity-service/src/main/java/uk/gov/di/ipv/core/library/service/UserIdentityService.java:51]: (and elsewhere)

Comment (DanCorderIPV) [reply libs/user-identity-service/src/main/java/uk/gov/di/ipv/core/library/service/UserIdentityService.java:51]: I didn't do this explicitly, not sure if it was spotless or IDEA but it seemed sensible. I can look into reversing it.

Comment (Joe-Edwards-GDS) [reply libs/user-identity-service/src/main/java/uk/gov/di/ipv/core/library/service/UserIdentityService.java:51]: Thanks - I expect it's IDEA (Mike had a similar thing)

File: libs/user-identity-service/src/main/java/uk/gov/di/ipv/core/library/service/UserIdentityService.java (hunk 2 above)
Comment (DanCorderIPV) [reply libs/user-identity-service/src/main/java/uk/gov/di/ipv/core/library/service/UserIdentityService.java:492]: I think it does make sense, the only difference is the error handling.

File: libs/user-identity-service/src/main/java/uk/gov/di/ipv/core/library/service/UserIdentityService.java [hunk 4]
```diff
@@ -333,131 +336,160 @@ public Optional<IdentityClaim> findIdentityClaim(List<VcStoreItem> vcStoreItems)
...
+
+    private Optional<VcStoreItem> findStoreItem(
+            List<String> criNames, List<VcStoreItem> vcStoreItems) {
+        return vcStoreItems.stream()
+                .filter(credential -> criNames.contains(credential.getCredentialIssuer()))
+                .findFirst();
+    }
+
+    private JsonNode extractCriNodeFromCredential(
+            String criName,
```
Comment (Joe-Edwards-GDS) [inline libs/user-identity-service/src/main/java/uk/gov/di/ipv/core/library/service/UserIdentityService.java]: Sorry, another small thing - this parameter isn't actually the CRI name, it's the property you're extracting. Perhaps `extractSubjectClaim()` or similar?

**Review by Joe-Edwards-GDS — DISMISSED** [review]

File: libs/user-identity-service/src/test/java/uk/gov/di/ipv/core/library/service/UserIdentityServiceTest.java [hunk 5]
```diff
@@ -274,9 +261,7 @@ void checkNameCorrelationWithMissingNameWithAddressCredentialsForReturnTrue() th
         List<VcStoreItem> vcStoreItems =
```
Comment (DanCorderIPV) [inline libs/user-identity-service/src/test/java/uk/gov/di/ipv/core/library/service/UserIdentityServiceTest.java:261]: I don't understand the name of this test so not sure what it is testing, but this is the change that was made on the main branch (removing the passport VC).

Comment (Joe-Edwards-GDS) [reply libs/user-identity-service/src/test/java/uk/gov/di/ipv/core/library/service/UserIdentityServiceTest.java:261]: I think the intent is that name correlation is happy to skip address credentials (which don't include a name)

**Review by Joe-Edwards-GDS — APPROVED** [review]

**Pull request merged**